controllers/apps/operations/upgrade_test.go

```diff
@@ -108,11 +273,164 @@ var _ = Describe("Upgrade OpsRequest", func() {
+						ServiceVersion:          pointer.String(serviceVer1),
+					},
+				},
+			})
+
+			By("expect for this opsRequest is Running and reuse the original ComponentDefinition")
+			reqCtx := intctrlutil.RequestCtx{Ctx: ctx}
+			makeUpgradeOpsIsRunning(reqCtx, opsRes)
+			Eventually(testapps.CheckObj(&testCtx, client.ObjectKeyFromObject(opsRes.Cluster), func(g Gomega, cluster *appsv1alpha1.Cluster) {
+				g.Expect(cluster.Spec.ComponentSpecs[0].ComponentDef).Should(Equal(compDef1.Name))
+				g.Expect(cluster.Spec.ComponentSpecs[0].ServiceVersion).Should(Equal(serviceVer1))
+			})).Should(Succeed())
+
+			By("expect upgrade successfully with the latest release of the specified serviceVersion")
+			mockPodsAppliedImage(release4)
+			expectOpsSucceed(reqCtx, opsRes, consensusComp)
+		})
+
+		It("Test upgrade OpsRequest when specified serviceVersion is empty", func() {
+			By("init operations resources")
+			compDef1, _, opsRes := initOpsResWithComponentDef(true)
+
+			By("create Upgrade Ops")
+			opsRes.OpsRequest = createUpgradeOpsRequest(opsRes.Cluster, appsv1alpha1.Upgrade{
+				Components: []appsv1alpha1.UpgradeComponent{
+					{
+						ComponentOps:            appsv1alpha1.ComponentOps{ComponentName: consensusComp},
+						ComponentDefinitionName: pointer.String(compDef1.Name),
+						ServiceVersion:          pointer.String(""),
+					},
+				},
+			})
+
+			By(" expect for this opsRequest is Running")
+			reqCtx := intctrlutil.RequestCtx{Ctx: ctx}
+			makeUpgradeOpsIsRunning(reqCtx, opsRes)
+			Eventually(testapps.CheckObj(&testCtx, client.ObjectKeyFromObject(opsRes.Cluster), func(g Gomega, cluster *appsv1alpha1.Cluster) {
+				g.Expect(cluster.Spec.ComponentSpecs[0].ComponentDef).Should(Equal(compDef1.Name))
+				g.Expect(cluster.Spec.ComponentSpecs[0].ServiceVersion).Should(BeEmpty())
+			})).Should(Succeed())
+
+			By("looking forward to using the latest serviceVersion and releaseVersion")
+			mockPodsAppliedImage(release4)
+			expectOpsSucceed(reqCtx, opsRes, consensusComp)
+		})
+
+		It("Test upgrade OpsRequest when serviceVersion is nil", func() {
+			By("init operations resources")
+			_, compDef2, opsRes := initOpsResWithComponentDef(true)
+
+			By("create Upgrade Ops")
+			opsRes.OpsRequest = createUpgradeOpsRequest(opsRes.Cluster, appsv1alpha1.Upgrade{
+				Components: []appsv1alpha1.UpgradeComponent{
+					{
+						ComponentOps:            appsv1alpha1.ComponentOps{ComponentName: consensusComp},
+						ComponentDefinitionName: pointer.String(compDef2.Name),
+					},
+				},
+			})
+
+			By("expecting no changes to serviceVersion")
+			reqCtx := intctrlutil.RequestCtx{Ctx: ctx}
+			makeUpgradeOpsIsRunning(reqCtx, opsRes)
+			Eventually(testapps.CheckObj(&testCtx, client.ObjectKeyFromObject(opsRes.Cluster), func(g Gomega, cluster *appsv1alpha1.Cluster) {
+				g.Expect(cluster.Spec.ComponentSpecs[0].ComponentDef).Should(Equal(compDef2.Name))
+				g.Expect(cluster.Spec.ComponentSpecs[0].ServiceVersion).Should(Equal(serviceVer0))
```

Comment:
should be "v2"?

Reply:
no, it will not update serviceVersion, if no specified serviceVersion(nil) rather than empty("")